### PR TITLE
feat/request-stack-voter

### PR DESCRIPTION
**Pull request by dominikzogg**

### CODE REVIEW

**Comment by dominikzogg** [comment]
Hello @stof 

What do you think about this PR?

Regards Dominik

**Comment by Nek-** [comment]
Hello,

Thanks for contributing. Can you tell me why the RouteVoter is not enough for you ?

Notice that the KnpMenuBundle sets up the route voter [for you](https://github.com/KnpLabs/KnpMenuBundle/blob/master/EventListener%2FVoterInitializerListener.php).

**Comment by dominikzogg** [comment]
@Nek cause there is no standalone Request within Symfony3 and Silex DIC anymore, see #222

**Comment by dbu** [comment]
i think this makes sense. the PR should deprecate RouteVoter and https://github.com/KnpLabs/KnpMenuBundle/blob/master/EventListener/VoterInitializerListener.php . and have some documentation and a changelog entry warning that the `request="true"` option on the knp_menu.voter tag will go away in the next major version.

maybe lets also add `@trigger_error` deprecation notices in https://github.com/KnpLabs/KnpMenuBundle/blob/5e7c44cb084b5c4f842f94ab443fb5dbb3f7a44c/DependencyInjection/Compiler/AddVotersPass.php#L32 so people realize they should not use that tag anymore.

File: src/Knp/Menu/Matcher/Voter/RequestStackRouteVoter.php [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+     * @param RequestStack $requestStack
+     */
+    public function __construct(RequestStack $requestStack)
+    {
+        $this->requestStack = $requestStack;
+    }
+
+    public function matchItem(ItemInterface $item)
+    {
+        if (null === $request = $this->requestStack->getCurrentRequest()) {
```
Comment (dbu) [inline src/Knp/Menu/Matcher/Voter/RequestStackRouteVoter.php]: should we have a config option whether to use current or master request? or two different classes for the 2 cases. the current implementation with the request listener only updates voters with the master request... so i would default to use the master request.

File: tests/Knp/Menu/Tests/Matcher/Voter/RequestStackRouteVoterTest.php [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+
+    /**
+     * @expectedException \InvalidArgumentException
+     */
+    public function testInvalidRouteConfig()
+    {
+        $item = $this->getMock('Knp\Menu\ItemInterface');
+        $item->expects($this->any())
+            ->method('getExtra')
+            ->with($this->logicalOr($this->equalTo('routes'), $this->equalTo('routesParameters')))
```
Comment (dbu) [inline tests/Knp/Menu/Tests/Matcher/Voter/RequestStackRouteVoterTest.php]: note that this just got simplified on the RouteVoterTest in master, see #244 - we should do the same here.

File: tests/Knp/Menu/Tests/Matcher/Voter/RequestStackRouteVoterTest.php [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                }
+            }));
+
+        $requestStack = new RequestStack();
+
+        $request = new Request();
+        $request->attributes->set('_route', 'foo');
+        $request->attributes->set('_route_params', array());
+
+        $requestStack->push($request);
```
Comment (dbu) [inline tests/Knp/Menu/Tests/Matcher/Voter/RequestStackRouteVoterTest.php]: we should have tests with more than one request in the stack, that make sure the right element on the stack is used (current resp master)

**Comment by stof** [comment]
We should update the RouterVoter class (maintaining BC) rather than creating a RequestStackRouteVoter. This is still a voter voting based on the route. the fact that it depends on the RequestStack is an implementation detail, which should not be in the name

**Comment by dominikzogg** [comment]
@dbu @stof i do not care which solution we or you choose, as long as it works and its not taking another half year to implement

**Comment by dbu** [comment]
@stof so after re-reading https://github.com/KnpLabs/KnpMenu/pull/222#issuecomment-141769541 your idea would be to have this new class, but call it RouteVoter as well and separate it by having it in the Integration\Symfony namespace instead of Matcher\Voter?

**Comment by dbu** [comment]
ping @stof

**Comment by dominikzogg** [comment]
@stof @dbu any news on this or alternative solution, which makes KnpMenu usable with Symfony 3?

**Comment by xabbuh** [comment]
You could change the existing voter to let it allow injecting either a `Request` or a `RequestStack` instance (making the decision what to inject in the container extension class in the bundle). Inside `matchItem()` you can then check what information is present.

**Comment by dominikzogg** [comment]
@xabbuh i see no reason to mix them, its more complex, less typesave

**Comment by stof** [comment]
@dominikzogg the idea is to have a single class, with a deprecated code path in it (which would trigger a deprecation warning when used).

**Comment by stof** [comment]
Closing as #222 was merged.

**Pull request closed without merging**